src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs

```diff
@@ -516,6 +550,82 @@ internal void ProcessDto(ModelBindingContext bindingContext, ComplexModelDto dto
             }
         }
 
+        // Neither [DefaultValue] nor [Required] is relevant for a non-settable collection.
+        private void AddToProperty(
+            ModelBindingContext bindingContext,
+            ModelExplorer modelExplorer,
+            PropertyInfo property,
+            ModelBindingResult dtoResult)
+        {
+            var propertyExplorer = modelExplorer.GetExplorerForProperty(property.Name);
+
+            var target = propertyExplorer.Model;
+            var source = dtoResult.Model;
+            if (!dtoResult.IsModelSet || target == null || source == null)
+            {
+                // Cannot copy to or from a null collection.
+                return;
+            }
+
+            // Determine T if this is an ICollection<T> property. No need for a T[] case because CanUpdateProperty()
+            // ensures property is either settable or not an array. Underlying assumption is that CanUpdateProperty()
+            // and SetProperty() are overridden together.
+            var elementTypeArray = propertyExplorer.ModelType
+                .ExtractGenericInterface(typeof(ICollection<>))
+                ?.GenericTypeArguments;
+            if (elementTypeArray == null)
+            {
+                // Not a collection model.
+                return;
+            }
+
+            var propertyAddRange = CallPropertyAddRangeOpenGenericMethod.MakeGenericMethod(elementTypeArray);
+            TryAction(
+                bindingContext,
+                dtoResult,
+                () => propertyAddRange.Invoke(obj: null, parameters: new[] { target, source }));
+        }
+
+        // Called via reflection.
+        private static void CallPropertyAddRange<TElement>(object target, object source)
+        {
+            var targetCollection = (ICollection<TElement>)target;
+            var sourceCollection = source as IEnumerable<TElement>;
+            if (sourceCollection != null && !targetCollection.IsReadOnly)
+            {
+                targetCollection.Clear();
+                foreach (var item in sourceCollection)
+                {
+                    targetCollection.Add(item);
+                }
+            }
+        }
+
+        private static void TryAction(ModelBindingContext bindingContext, ModelBindingResult dtoResult, Action action)
+        {
+            try
+            {
+                action();
```

Comment:
I'm really not a big fan of this pattern because it makes the debugging experience pretty poor.  How would you feel about factoring out the code inside the catch block to a private method instead?